### PR TITLE
Diamond basic block inlining for IFELSE

### PR DESCRIPTION
**Pull request by LenyKholodov**
This PR provides IFELSE optimizations on peephole stage:
- inlining post-dominant BB after IFELSE instruction instead of JMPX to it from THEN/ELSE branches;
- removing of empty PUSHCONT {} branches and replacing them to IF or IFNOT.

### CODE REVIEW

File: llvm/lib/Target/TVM/TVMComparisonInstrInfo.td [hunk 1]
```diff
@@ -82,6 +82,10 @@ defm SDCNTTRAIL1 : I<(outs I257:$cnttrail1), (ins Slice:$sl), (outs), (ins),
                      "SDCNTTRAIL1\t$cnttrail1, $sl", "SDCNTTRAIL1", 0xc713>;
 
 // A.3.1. Null primitives
+defm PUSHNULL : I0<(outs I257:$result), (ins),
+                 [],
```
Comment (akiramenai) [inline llvm/lib/Target/TVM/TVMComparisonInstrInfo.td]: nitpicking: misindented

File: llvm/lib/Target/TVM/TVMControlFlowInstrInfo.td [hunk 2]
```diff
@@ -38,6 +38,15 @@ let isTerminator = 1, hasCtrlDep = 1 in {
...
                    [(TVMifelse I257:$cond, I257:$branch1, I257:$branch2)],
                     "IFELSE\t$branch1, $branch2, $cond", "IFELSE", 0xe2>;
+
+    defm IF : I<(outs), (ins I257:$cond, I257:$dst),
+                (outs), (ins),
+                [(TVMif I257:$cond, I257:$dst)],
+                "IF\t$dst, $cond", "IF", 0xde>;
+    defm IFNOT : I<(outs), (ins I257:$cond, I257:$dst),
+                   (outs), (ins),
+                   [(TVMifnot I257:$cond, I257:$dst)],
```
Comment (akiramenai) [inline llvm/lib/Target/TVM/TVMControlFlowInstrInfo.td]: Do we really need to have that instead of `TVMif(not ...)`?

Comment (LenyKholodov) [reply llvm/lib/Target/TVM/TVMControlFlowInstrInfo.td]: Fixed

File: llvm/lib/Target/TVM/TVMComparisonInstrInfo.td (hunk 1 above)
Comment (LenyKholodov) [reply llvm/lib/Target/TVM/TVMComparisonInstrInfo.td]: Fixed

File: llvm/lib/Target/TVM/TVMStackModel.cpp [hunk 3]
```diff
@@ -685,6 +685,17 @@ void TVMStackModel::rewriteToSForm(MachineInstr &MI, std::string &PreTermStackSt
       }
     }
 
+    std::vector<unsigned> SourceRegs;
+
+    for (unsigned I = 0; I < MI.getNumOperands(); I++) {
```
Comment (akiramenai) [inline llvm/lib/Target/TVM/TVMStackModel.cpp]: `std::copy_if`?

File: llvm/lib/Target/TVM/TVMMachineFunctionInfo.h [hunk 4]
```diff
@@ -128,6 +133,41 @@ class TVMFunctionInfo final : public MachineFunctionInfo {
...
+  }
+  const std::vector<unsigned>* getStackModelSourceRegs(const MachineInstr *Target) {
+    auto It = StackModelSourceRegs.find(Target);
+    if (It == StackModelSourceRegs.end())
+      return nullptr;
+    return &It->second;
+  }
+
+  // Clone machine instruction intermediate data
+  void cloneMachineInstrIntermediateData(const MachineInstr *Source, const MachineInstr *Destination) {
```
[the remark this answers is not in the document]
Comment (LenyKholodov) [reply llvm/lib/Target/TVM/TVMMachineFunctionInfo.h]: Fixed

File: llvm/lib/Target/TVM/TVMMachineFunctionInfo.h [hunk 5]
```diff
@@ -128,6 +133,41 @@ class TVMFunctionInfo final : public MachineFunctionInfo {
...
+    StackModelComments[Destination].clear();
+    for (const auto& Comment : getStackModelComments(Source)) {
+      addStackModelComment(Destination, Comment);
+    }
+
+    if (const auto* Regs = getStackModelSourceRegs(Source)) {
+      std::vector<unsigned> RegsCopy (*Regs);
+      setStackModelSourceRegs(Destination, RegsCopy);
```
[the remark this answers is not in the document]
Comment (LenyKholodov) [reply llvm/lib/Target/TVM/TVMMachineFunctionInfo.h]: Fixed

File: llvm/lib/Target/TVM/TVMMachineFunctionInfo.h [hunk 6]
```diff
@@ -128,6 +133,41 @@ class TVMFunctionInfo final : public MachineFunctionInfo {
...
+    }
+  }
+
+  // Clear MachineInstr intermediate data
+  const MachineInstrBuilder& clearIntermediateData(const MachineInstrBuilder &Builder) {
+    clearIntermediateData(Builder.getInstr());
+    return Builder;
+  }
```
[the remark this answers is not in the document]
Comment (LenyKholodov) [reply llvm/lib/Target/TVM/TVMMachineFunctionInfo.h]: Fixed

File: llvm/lib/Target/TVM/TVMPeephole.cpp [hunk 7]
```diff
@@ -99,7 +108,27 @@ bool TVMPeephole::runMbbInlineOptimization(MachineBasicBlock &MBB) {
 
   ++InstrIter;
 
-  if (InstrIter == MBB.instr_rend())
+  auto FirstStackInstrIter = InstrIter;
+  bool PushContMBBFound = false;
+
+  for (; InstrIter != MBB.instr_rend() && !PushContMBBFound;) {
```
[the remark this answers is not in the document]
Comment (LenyKholodov) [reply llvm/lib/Target/TVM/TVMPeephole.cpp]: Fixed

File: llvm/lib/Target/TVM/TVMPeephole.cpp [hunk 8]
```diff
@@ -99,7 +108,27 @@ bool TVMPeephole::runMbbInlineOptimization(MachineBasicBlock &MBB) {
...
+    switch (InstrIter->getOpcode()) {
```
[the remark this answers is not in the document]
Comment (LenyKholodov) [reply llvm/lib/Target/TVM/TVMPeephole.cpp]: This loop processes only the sequence of NIP and one PUSHCONT_MBB. Other instructions may have different transformations (at this moment NIP transforms to DROP, and this is not general case; so for other sequences pattern match fails and optimization is disabled).

File: llvm/lib/Target/TVM/TVMPeephole.cpp [hunk 9]
```diff
@@ -117,9 +146,29 @@ bool TVMPeephole::runMbbInlineOptimization(MachineBasicBlock &MBB) {
   const MachineBasicBlock &SourceMBB = *PushContMBB.getOperand(0).getMBB();
   auto InsertionIter = InstrIter.getReverse();
   MachineFunction *MF = MBB.getParent();
+  TVMFunctionInfo *MFI = MF->getInfo<TVMFunctionInfo>();
+
+  // Convert MI sequence to expected form (PUSHCONT, JMPX) and move all
+  // stack operations between PUSHCONT and JMPX before the PUSHCONT
+  // (the list of possible transformations may be extended in future)
+  for (InstrIter = FirstStackInstrIter; InstrIter != LastStackInstrIter;) {
+    switch (InstrIter->getOpcode()) {
```
[the remark this answers is not in the document]
Comment (LenyKholodov) [reply llvm/lib/Target/TVM/TVMPeephole.cpp]: Fixed

File: llvm/lib/Target/TVM/TVMStackModel.cpp (hunk 3 above)
Comment (LenyKholodov) [reply llvm/lib/Target/TVM/TVMStackModel.cpp]: This is not pure copy_if. In this case we need something like transform_if, which is not present in STL. That's why the transformation loop has been written manually.

**Comment by LenyKholodov** [comment]
> Why this optimization should work after stack modelling? Not with virtual registers state of machine code?

This patch provides simple solution for optimization of common code patterns (IFELSE pattern is used very frequently). Also, it provides MachineInstrMatcher which may be used for further peephole transformations. General case should be definitely resolved using analyzeBranch/removeBranch and partially before stack modelling. So this PR provides solution which does not conflict with general case implementation and may be used before it appears.

File: llvm/lib/Target/TVM/TVMPeephole.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (akiramenai) [reply llvm/lib/Target/TVM/TVMPeephole.cpp]: I didn't get it. What is the difference between the case and:
```
auto It = find_if(InstrIter, MBB.instr_rend(), [](auto& Instr) {Instr.getOpcode() != TVM::NIP;});
if (It == MBB.instr_rend() || It->getOpcode() != TVM::PUSHCONT_MBB_S)
  return false;
```

If there is no difference, I think that `find_if` expresses the intent better than `switch`, doesn't it?

File: llvm/lib/Target/TVM/TVMStackModel.cpp (hunk 3 above)
Comment (akiramenai) [reply llvm/lib/Target/TVM/TVMStackModel.cpp]: Ok, got it. Didn't notice `getReg()`. Then let's not recalculate `MI.getNumOperands()`.

**Pull request merged**